content/blog/2016/2016-12-13-jenkins-puppet-enterprise-plugin.adoc

```diff
@@ -0,0 +1,271 @@
+
+----
+/opt/puppetlabs/puppet/bin/gem install hiera-http
+/opt/puppetlabs/bin/puppetserver gem install hiera-http
+----
+
+Now you can configure the /etc/puppetlabs/puppet/hiera.yaml file. The following
+configuration instructs Hiera to first look to the Hiera yaml files in the
+Puppet code's environment, then fall back to the http backend. The http backend
+will first query the Hiera data store API looking for the key in the scope with
+the same name as the node. If nothing's found, look for the key in the node's
+environments. You can use any Facter fact to match scope names.
+
+----
+:backends:
+  - yaml
+  - http
+
+:http:
+  :host: jenkins.example.com
+  :port: 8080
+  :output: json
+  :use_auth: true
+  :auth_user: <user>
+  :auth_pass: <pass>
+  :cache_timeout: 10
+  :failure: graceful
+  :paths:
+    - /hiera/lookup?path=%{clientcert}&key=%{key}
+    - /hiera/lookup?path=%{environment}&key=%{key}
+----
+
+Finally, restart the pe-puppetserver process to pick up the new configs.
+----
+/opt/puppetlabs/bin/puppet resource service pe-puppetserver ensure=stopped
+/opt/puppetlabs/bin/puppet resource service pe-puppetserver ensure=running
+----
+
+==== Hiera HTTP Authentication
+
+If Jenkins' Global Security is configured to allow unauthenticated read-only
+access, the 'use_auth', 'auth_pass', and 'auth_user' parameters are
+unnecessary. Otherwise, create a local Jenkins user that has permissions to
+view the Hiera Data Lookup page and use that user's credentials for the
+hiera.yaml configuration.
+
+== Querying the infrastructure
+
+PuppetDB is an extensive data store that holds every bit of information Puppet
+generates and collects across every system Puppet is installed on. PuppetDB's
+provides a sweet query language called
+link:https://docs.puppet.com/puppetdb/4.3/api/query/v4/pql.html[PQL.] With PQL,
+you can ask complex questions of your infrastructure such as "How production
+Red Hat systems are there with the openssl package installed?" or "What are are
+the us-west-2c nodes with the MyApp role that were created in the last 24
+hours?"
+
+This can be a pretty powerful tool for parts of your pipeline where you need to
+perform specific operations on subsets of the infrastructure like draining a
+loadbalancer.
+
+To perform a query, use the puppet.query method.
+
+----
+results = puppet.query 'inventory[certname] { facts.os.name = "RedHat" and facts.ec2_metadata.placement.availability-zone = "us-west-2c" and uptime_hours < 24 }'
+----
+
+The value returned is an array of items matching the query. The results can be
+iterated on, and even passed to a series of puppet.job calls. For example, the
+following code will query all nodes in production that experienced a failure on
+the last Puppet run, break the resulting list into groups of 3, and runs Puppet
+on each sub group.
+
+----
+results = puppet.query 'nodes { latest_report_status = "failed" and catalog_environment = "production"}'
+nodes = []
+for (Map node : results) {
```

Comment:
https://github.com/jenkinsci/pipeline-examples/blob/master/docs/BEST_PRACTICES.md#groovy-gotchas 
> Beware for (Foo f: foos) loops and Groovy closure-style operators like .each and the like. They will not work right in normal Pipeline script contexts where Pipeline steps are involved directly.

This might be fine but it is unclear from where this code would run.   It is likely to be copied by others and fail in odd ways.